core/upgrade/src/main/java/org/opennms/upgrade/implementations/RemotePollerServiceConfigMigratorOffline.java

```diff
@@ -0,0 +1,177 @@
+import java.io.StringWriter;
+import java.util.Collections;
+
+import org.apache.commons.io.FileUtils;
+import org.opennms.core.utils.ConfigFileConstants;
+import org.opennms.core.xml.JaxbUtils;
+import org.opennms.netmgt.config.service.Attribute;
+import org.opennms.netmgt.config.service.Invoke;
+import org.opennms.netmgt.config.service.InvokeAtType;
+import org.opennms.netmgt.config.service.Service;
+import org.opennms.netmgt.config.service.ServiceConfiguration;
+import org.opennms.upgrade.api.AbstractOnmsUpgrade;
+import org.opennms.upgrade.api.OnmsUpgradeException;
+
+import com.google.common.collect.Lists;
+
+public class RemotePollerServiceConfigMigratorOffline extends AbstractOnmsUpgrade {
+
+    private File configFile;
+
+    public static final String DEPRECATED_REMOTE_POLLER_SERVICENAME = "OpenNMS:Name=PollerBackEnd";
+    public static final String REMOTE_POLLER_NG_SERVICENAME = "OpenNMS:Name=RemotePollerNG";
+
+    public RemotePollerServiceConfigMigratorOffline() throws OnmsUpgradeException {
+        super();
+        try {
+            configFile = ConfigFileConstants.getFile(ConfigFileConstants.SERVICE_CONF_FILE_NAME);
+        } catch (final IOException e) {
+            throw new OnmsUpgradeException("Can't find Services Configuration file", e);
+        }
+    }
+
+    @Override
+    public int getOrder() {
+        return 14;
+    }
+
+    @Override
+    public String getDescription() {
+        return "Remove deprecated RemotePoller service entry from service-configuration.xml, see NMS-12684";
+    }
+
+    @Override
+    public void preExecute() throws OnmsUpgradeException {
+        try {
+            log("Creating backup of %s\n", configFile);
+            zipFile(configFile);
+        } catch (Exception e) {
+            throw new OnmsUpgradeException("Can't backup service-configurations.xml because " + e.getMessage());
+        }
+
+    }
+
+    @Override
+    public void postExecute() throws OnmsUpgradeException {
+        File zip = new File(configFile.getAbsolutePath() + ZIP_EXT);
+        if (zip.exists()) {
+            log("Removing backup %s\n", zip);
+            FileUtils.deleteQuietly(zip);
+        }
+    }
+
+    @Override
+    public void rollback() throws OnmsUpgradeException {
+        log("Restoring backup %s\n", configFile);
+        File zip = new File(configFile.getAbsolutePath() + ZIP_EXT);
+        FileUtils.deleteQuietly(configFile);
+        unzipFile(zip, zip.getParentFile());
+    }
+
+    @Override
+    public void execute() throws OnmsUpgradeException {
+        try {
+            final ServiceConfiguration currentCfg = JaxbUtils.unmarshal(ServiceConfiguration.class, configFile);
+            boolean skipRemovePollerNgEntryCreation = false;
+            boolean deprecatedServiceEnabled = true;
+
+            log("Current configuration: " + currentCfg.getServices().size() + " services.\n");
+
+            for (int i = currentCfg.getServices().size() - 1; i >= 0; i--) {
```

Comment:
`for (final Service localSvc : Lists.reverse(currentCfg.getServices())) {` looks more modern